### PR TITLE
Optimizer Improvements and Stochastic weight averaging

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
removing ranger21 in favour of stripped down custom implementation adding rangerlite optimizer option with a bugfix (see https://github.com/lessw2020/Ranger21/pull/54 for details)

To verify yourself that the lite version does the same see https://github.com/TonyCongqianWang/RangerLite.

All in all it is a bit faster than master on nettest hardware.

Adding options to training loss, lr scheduling and also stochastic weight averaging. Allowing for more extensive hyperparameter tuning.

The most impactful change is the introduction of lambda pertubation parameters. They allow a training recipe without any predefined lambda scheduling which greatly improves the ability to tune hyperparameters without them being highly dependent on training length.

### CODE REVIEW

File: model/optimizers/schedulefree_wrapper.py [hunk 1]
```diff
@@ -19,48 +19,50 @@ class ScheduleFreeConfig:
...
+
     def configure_optimizers(self, train_params):
         if _schedulefree_import_error:
             raise ImportError("The required schedulefree library is not installed. ")
         print(
-            f"Using schedule-free Adam with warmup_steps={self.warmup_steps}, lr={self.lr}."
+            f"[SchedulefreeSetup] warmup_steps={self.warmup_steps}."
         )
-        optimizer = schedulefree.AdamWScheduleFree(
+        self.optimzier = schedulefree.AdamWScheduleFree(
```
Comment (vondele) [inline model/optimizers/schedulefree_wrapper.py]: typo?

File: model/optimizers/schedulefree_wrapper.py [hunk 2]
```diff
@@ -19,48 +19,50 @@ class ScheduleFreeConfig:
...
             train_params,
             lr=self.lr,
             betas=(0.9, 0.999),
             eps=1.0e-7,
             warmup_steps=self.warmup_steps,
         )
-        return optimizer
+        return self.optimzier
```
Comment (vondele) [inline model/optimizers/schedulefree_wrapper.py]: typo

**Comment by TonyCongqianWang** [comment]
I accidentally also merged in #430 so it has to be either cleaned up before or we merge #430 first (though it still needs to be shown to be non regression)

File: model/callbacks.py [hunk 3]
```diff
@@ -15,3 +17,31 @@ def on_train_batch_start(
...
+        self.swa_model.update_parameters(pl_module.model)
+        pl_module.train()
+
+    def state_dict(self):
+        # Prevent Lightning from saving the SWA callback's internal state
+        # in the regular epoch checkpoints, avoiding redundant memory bloat.
+        return {}
+
+    def load_state_dict(self, state_dict):
+        pass
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/callbacks.py:50]: Good point, though supporting resume from ckpt with swa is not worth the trouble

[the remark this answers is not in the document]
Comment (vondele) [reply model/callbacks.py:50]: would it make sense to add a kind of check/assert there to ensure the code aborts in this case.

[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/callbacks.py:50]: Hmmm perhaps.. not trivial though as the swa callback is agnostic to the current epoch and training state in general...

Or actually the easiest is to check in train.py if the resumed epoch is `>= swa_start_epoch` and abort in that case.. that should work

**Comment by vondele** [comment]
we should probably turn this branch into a mergable PR that matches the code as needed to generate the currently pending https://github.com/official-stockfish/Stockfish/pull/6802

File: train.py [hunk 4]
```diff
@@ -453,24 +436,32 @@ def main():
...
+            checkpoint_callback,
+            SimpleLineLogger(refresh_rate=refresh_rate),
+            TimeLimitAfterCheckpoint(args.max_time),
+            M.WeightClippingCallback(),
+        ]
+    if 0 <= args.swa_start_epoch < args.max_epochs:
+        swa_callback = M.ExplicitSWACallback(args.swa_start_epoch)
+        trainer_callbacks.append(
+            swa_callback
+        )
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply train.py:463]: It does explicitly throw an error in the callback

File: train.py [hunk 5]
```diff
@@ -481,6 +472,34 @@ def main():
...
+        # Use train mode for lightning module but eval mode for model,
+        # prevents backup params from bloating ckpt.
+        # Note that resume from checkpoint after swa averaging has started.
+        nnue.train()
+        nnue.model.eval()
+        nnue.model.load_state_dict(swa_state_dict)
+
+        # NOTE: If BN is used, it has to be updated here. Be careful when using DDP.
+        swa_savepath = os.path.join(logdir, "lightning_logs", f"version_{tb_logger.version}", "checkpoints", "last_swa.ckpt")
+        trainer.save_checkpoint(swa_savepath)
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply train.py]: Would be nice, but unfortunately not supported with current checkpointing resuming logic. Would be a worthwhile refactor though

File: model/lightning_module.py [hunk 6]
```diff
@@ -207,40 +291,46 @@ def step_(self, batch: tuple[Tensor, ...], batch_idx, loss_type):
...
+            reg_loss = ft_act_loss(ft_activation, loss_params.ft_activation_l1, loss_params.ft_activation_l2)
+        else:
+            reg_loss = torch.zeros_like(fit_loss)
+        loss = fit_loss + reg_loss
+
+        self.loss_metrics[f"{loss_type}_epoch"].update(fit_loss)
         self.log(
             loss_type,
-            loss,
+            fit_loss,
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/lightning_module.py:308]: I will remove the unused regularization loss for now.

It would be best to log all separately when adding auxiliary losses as the sum isn't comparable

File: model/utils/serialize.py [hunk 7]
```diff
@@ -79,6 +79,9 @@ def __init__(
 
         self.buf = bytearray()
 
+        model.clip_weights()
+        model.clip_input_weights()
+
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/utils/serialize.py:84]: I am aware of this, it is something I am working to refactor, but there is no easy fix to ensure threat weights are in a safe range

File: model/optimizers/ranger_lite.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        if not self.lookahead_active:
+            return
+        for group in self.param_groups:
+            for p in group["params"]:
+                state = self.state[p]
+                if "lookahead_params" in state:
+                    # Only backup if we haven't already backed up
+                    if "backup_params" not in state:
+                        state["backup_params"] = torch.clone(p.data)
+                    p.data.copy_(state["lookahead_params"])
```
Comment (xu-shawn) [inline model/optimizers/ranger_lite.py]: Claude Ultrareview

`model/optimizers/ranger_lite.py:282` — swap_for_inference is not idempotent → corrupts SWA save and validation
The `p.data.copy_(state['lookahead_params'])` runs unconditionally even when `backup_params` already exists. During SWA finalization, Lightning captures the `state_dict` (aliasing live tensors) then fires `on_save_checkpoint` which calls `swap_for_inference`, overwriting SWA weights with lookahead weights in-place — so `last_swa.ckpt` contains lookahead, not SWA. Fix: gate the `p.data.copy_` inside the same `if "backup_params"` not in state block.

File: model/optimizers/ranger_lite.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+                variance_ma_debiased = variance_ma / bias_correction2
+                variance_ma_sum += variance_ma_debiased.sum()
+
+        if not self.param_size:
+            if not param_size:
+                # No trainable params
+                return loss
+            self.param_size = param_size
+
+        variance_normalized = torch.sqrt(variance_ma_sum / self.param_size)
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/optimizers/ranger_lite.py]: This is not done in ranger21 so I am a bit hesitant to add this although clamping to a very small value shouldn't do anything but in extreme edge cases

File: model/optimizers/ranger_lite.py (hunk 8 above)
Comment (TonyCongqianWang) [reply model/optimizers/ranger_lite.py]: Yeah... I thought I had already fixed that by using `.eval()` but no... Copilot also flagged that

Comment (TonyCongqianWang) [reply model/optimizers/ranger_lite.py]: Lookahead's leaky abstraction giving some headaches....

**Comment by TonyCongqianWang** [comment]
no Idea why I used copies in the first place... got lead astray by Gemini probably....

Changed that a bunch of times back and forth with Copilot suggestion etc, etc.. probably made it worse at some point

Hopefully everything fixed now...

**Comment by TonyCongqianWang** [comment]
Was merged with separate PRs
